core/src/vaults.ts

```diff
@@ -0,0 +1,446 @@
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 1,
+});
+
+const _fetchCollateralLiquidationLimitsAndLiquidatorAddress = async (
+    network: string,
+    collateralType: CollateralType
+) => {
+    const contract = await getContract(network, 'MCD_DOG');
+    const contractVat = await getContract(network, 'MCD_VAT');
+    const typeHex = ethers.utils.formatBytes32String(collateralType);
+    const { hole, dirt, clip, chop } = await contract.ilks(typeHex);
+    const { dust } = await contractVat.ilks(typeHex);
+    return {
+        currentCollateralDebtDai: new BigNumber(dirt._hex).shiftedBy(-RAD_NUMBER_OF_DIGITS),
+        maximumCollateralDebtDai: new BigNumber(hole._hex).shiftedBy(-RAD_NUMBER_OF_DIGITS),
+        liquidatiorContractAddress: clip,
+        liquidationPenaltyRatio: new BigNumber(chop._hex).shiftedBy(-WAD_NUMBER_OF_DIGITS),
+        minimalAuctionedDai: new BigNumber(dust._hex).shiftedBy(-RAD_NUMBER_OF_DIGITS),
+    };
+};
+
+export const fetchCollateralLiquidationLimitsAndLiquidatorAddress = memoizee(
+    _fetchCollateralLiquidationLimitsAndLiquidatorAddress,
+    {
+        maxAge: CACHE_EXPIRY_MS,
+        promise: true,
+        length: 2,
+    }
+);
+
+const _fetchVault = async (network: string, index: number): Promise<Vault> => {
+    const vaultBase = await fetchVaultBase(network, index);
+    const vaultCollateralParameters = await fetchVaultCollateralParameters(network, vaultBase.collateralType);
+    const vaultAmount = await fetchVaultAmount(network, vaultBase.collateralType, vaultBase.address);
+    const globalLiquidationLimits = await fetchGlobalLiquidationLimits(network);
+    const { currentCollateralDebtDai, maximumCollateralDebtDai, liquidationPenaltyRatio, minimalAuctionedDai } =
+        await fetchCollateralLiquidationLimitsAndLiquidatorAddress(network, vaultBase.collateralType);
+    return {
+        ...vaultBase,
+        ...vaultCollateralParameters,
+        ...vaultAmount,
+        ...globalLiquidationLimits,
+        currentCollateralDebtDai,
+        maximumCollateralDebtDai,
+        liquidationPenaltyRatio,
+        minimalAuctionedDai,
+    };
+};
+export const fetchVault = memoizee(_fetchVault, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 2,
+});
+
+const getNextOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+): Promise<BigNumber> => {
+    if (!('nextPriceSlotAddress' in oracle)) {
+        return new BigNumber(NaN);
+    }
+    const slot = oracle.nextPriceSlotAddress;
+    const nextPriceFeed = await provider.getStorageAt(oracleAddress, slot);
+    const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+    const isPriceValid = nextPriceFeed.substring(0, valueSplitPosition);
+    if (parseInt(isPriceValid, 16) === 1) {
+        return new BigNumber(`0x${nextPriceFeed.substring(valueSplitPosition)}`);
+    }
+    return new BigNumber(NaN);
```

Comment:
As [previously suggested](https://github.com/sidestream-tech/unified-auctions-ui/pull/457#discussion_r971857568): 
- [ ] please add a comment in the code explaining what's happening here (for me it looks like a bunch of hacks with no meaning)
- [ ] please add section to the "collateral onboarding" guide to explain how to determine which collateral type it is and where to get those values 
- [ ] please provide a method to validate that those values are still valid (eg right now it is tested on a specific block, but how to check if those are still correct if when we run it on the latest block?)

Reply:
> but how to check if those are still correct if when we run it on the latest block?)

i don't think this one is feasible. Since we receive bytes as the result of memory access, we cannot be sure that the accessed value is still the one we're looking for. The only case where we can likely say that the logic is broken is to fetch 0 value (since price 0 is not possible???).

Reply:
So the only option is to manually ballpark received number or how exactly do you manually validate it?

Then, should we record the address of the oracle in the config to make sure it hasn't changed? And make a script that alerts user if the contract is different (eg in the CI)?